stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                    "parent_tenure" => %sortition_state.parent_tenure_id,
+                    "last_sortition" => %sortition_state.prior_sortition,
+                    "violating_tenure_id" => %tenure.consensus_hash,
+                    "violating_tenure_first_block_id" => ?tenure.first_block_mined,
+                );
+                return Ok(false);
+            }
+        }
+
+        return Ok(true);
+    }
+
+    fn check_tenure_change_block_confirmation(
+        tenure_change: &TenureChangePayload,
+        block: &NakamotoBlock,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        // in tenure changes, we need to check:
+        // (1) if the tenure change confirms the expected parent block (i.e.,
+        //     the last block we signed in the parent tenure)
+        // (2) if the parent tenure was a valid choice
+        Self::confirms_known_blocks_in(block, &tenure_change.prev_tenure_consensus_hash, signer_db)
+    }
+
+    fn confirms_known_blocks_in(
+        block: &NakamotoBlock,
+        tenure: &ConsensusHash,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        let Some(last_known_block) = signer_db
+            .get_last_signed_block_in_tenure(tenure)
+            .map_err(|e| ClientError::InvalidResponse(e.to_string()))?
+        else {
+            info!(
+                "Have not signed off on any blocks in the parent tenure, assuming block confirmation is correct";
+                "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                "tenure" => %tenure,
+            );
+            return Ok(true);
+        };
+        if block.header.chain_length > last_known_block.block.header.chain_length {
+            return Ok(true);
+        } else {
+            warn!(
+                "Miner block proposal's tenure change transaction does not confirm as many blocks as we expect in the parent tenure";
+                "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                "proposed_chain_length" => block.header.chain_length,
+                "expected_at_least" => last_known_block.block.header.chain_length + 1,
+            );
+            return Ok(false);
+        }
+    }
+
+    /// Has the current tenure lasted long enough to extend the block limit?
+    pub fn tenure_time_passed_block_lim() -> Result<bool, ClientError> {
+        // TODO
+        return Ok(false);
+    }
+
+    /// If necessary, fetch a new view of the recent sortitions
+    pub fn refresh_view(&mut self, client: &StacksClient) -> Result<(), ClientError> {
```

Comment:
This function (or something like it) is going to need to be written to _recursively_ search the node's sortition history for the highest common ancestor that is present in both the node and the local signer DB, and _correct_ the signer DB so that its history of signed Stacks blocks corresponds to the node's sortition history.  This is necessary to keep the signer DB consistent with the node's view of the canonical Bitcoin fork.  Stacks blocks in the signer DB are going to need to be marked as non-canonical (and thus not worthy of consideration) in the event that their miner's sortition gets invalidated through a Bitcoin reorg.

Reply:
I don’t think so— the signer db stores blocks signed for given tenure IDs, which are unique to each bitcoin fork. As long as the signer is querying using the canonical sortition tip (which is what the new RPC endpoints are trying to enable it to do), specifically tracking canonical status in the signerdb shouldn’t be necessary.